File: OpenDiablo2.Scenes/Credits.cs
using OpenDiablo2.Common;
using OpenDiablo2.Common.Attributes;
using OpenDiablo2.Common.Enums;
using OpenDiablo2.Common.Interfaces;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OpenDiablo2.Scenes
{
    [Scene(eSceneType.Credits)]
    public sealed class Credits : IScene
    {
        private class LabelItem
        {
            public ILabel Label { get; set; }
            public bool IsHeading { get; set; }
            public bool Avaiable { get; set; }
        }

        private const float secondsPerCycle = (40f / 1000f);

        private readonly IRenderWindow renderWindow;
        private readonly ISceneManager sceneManager;
        private readonly IMPQProvider mpqProvider;

        private bool doneWithCredits = false;
        private int cyclesTillNextLine = 0;
        private float cycleTime = 0f;
        private IFont textFont;
        private ISprite backgroundSprite;
        private IButton btnExit;
        private Stack<string> creditsText;
        private List<LabelItem> fontLabels = new List<LabelItem>();

        public Credits(
            IRenderWindow renderWindow,
            ISceneManager sceneManager,
            IMPQProvider mpqProvider,
            Func<eButtonType, IButton> createButton
            )
        {
            this.renderWindow = renderWindow;
            this.sceneManager = sceneManager;
            this.mpqProvider = mpqProvider;

            backgroundSprite = renderWindow.LoadSprite(ResourcePaths.CreditsBackground, Palettes.Sky);

            btnExit = createButton(eButtonType.Medium);
            btnExit.Text = "Exit".ToUpper();
            btnExit.Location = new Point(20, 550);
            btnExit.OnActivate = OnActivateClicked;

            textFont = renderWindow.LoadFont(ResourcePaths.FontFormal10, Palettes.Static);

            creditsText = new Stack<string>((
                File.ReadAllLines(Path.Combine(Path.GetDirectoryName(Environment.GetCommandLineArgs().First()), "credits.txt"))
                    .Concat(mpqProvider.GetTextFile(ResourcePaths.CreditsText))
                ).Reverse());
        }

        private void OnActivateClicked()
            => sceneManager.ChangeScene(eSceneType.MainMenu);

        private void AddNextItem()
        {
            if (!creditsText.Any())
            {
                doneWithCredits = true;
                return;
            }

            var text = creditsText.Pop().Trim();
            if (text.Trim().Length == 0)
            {
                cyclesTillNextLine = 18;
                return;
            }
            var isHeading = text.StartsWith("*");
            var isNextHeading = creditsText.Any() && creditsText.Peek().StartsWith("*");
            var isNextSpace = creditsText.Any() && creditsText.Peek().Trim().Length == 0;
            var label = GetNewFontLabel(isHeading);
            label.Text = isHeading ? text.Substring(1) : text;
            var isDoubled = false;
            if (!isHeading && !isNextHeading && !isNextSpace)
            {
                isDoubled = true;

                // Gotta go side by side
                label.Location = new Point(390 - label.TextArea.Width, 605);

                var text2 = creditsText.Pop().Trim();
                isNextHeading = creditsText.Any() && creditsText.Peek().StartsWith("*");
                var label2 = GetNewFontLabel(isHeading);
                label2.Text = text2;

                label2.Location = new Point(410, 605);
            }
            else
            {
                label.Location = new Point(400 - (label.TextArea.Width / 2), 605);
            }

            if (isHeading && isNextHeading)
                cyclesTillNextLine = 40;
            else if (isNextHeading)
                cyclesTillNextLine = isDoubled ? 40 : 70;
            else if (isHeading)
                cyclesTillNextLine = 40;
            else
                cyclesTillNextLine = 18;
        }

        public void Render()
        {
            renderWindow.Draw(backgroundSprite, 4, 3, 0);
            btnExit.Render();
            foreach (var label in fontLabels.Where(x => !x.Avaiable).Select(x => x.Label))
                renderWindow.Draw(label);
        }

        public void Update(long ms)
        {

            cycleTime += (ms / 1000f);
            while (cycleTime >= secondsPerCycle)
            {
                cycleTime -= secondsPerCycle;
                if (!doneWithCredits && (--cyclesTillNextLine <= 0))
                    AddNextItem();

                foreach (var fontLabel in fontLabels.Where(x => !x.Avaiable))
                {
                    if (fontLabel.Label.Location.Y - 1 <= -15)
                    {
                        fontLabel.Avaiable = true;
                        continue;
                    }
                    fontLabel.Label.Location = new Point(fontLabel.Label.Location.X, fontLabel.Label.Location.Y - 1);
                }
            }

            btnExit.Update();
        }

        private ILabel GetNewFontLabel(bool isHeading)
        {
            var labelItem = fontLabels.FirstOrDefault(x => x.Avaiable && x.IsHeading == isHeading);
            if (labelItem != null)
            {
                labelItem.Avaiable = false;
                return labelItem.Label;
            }

            var newLabelItem = new LabelItem
            {
                Avaiable = false,
                IsHeading = isHeading,
                Label = renderWindow.CreateLabel(textFont)
            };

            newLabelItem.Label.TextColor = isHeading
                ? Color.FromArgb(255, 88, 82)
                : Color.FromArgb(198, 178, 150);

            fontLabels.Add(newLabelItem);
            return newLabelItem.Label;
        }


        public void Dispose()
        {
            backgroundSprite?.Dispose();

            foreach (var labelItem in fontLabels)
                labelItem.Label.Dispose();

            textFont.Dispose();
        }
    }
}

File: scenes/Credits.go
package scenes

import (
	"image/color"
	"strings"

	"github.com/OpenDiablo2/OpenDiablo2/common"
	"github.com/OpenDiablo2/OpenDiablo2/palettedefs"
	"github.com/OpenDiablo2/OpenDiablo2/resourcepaths"
	"github.com/OpenDiablo2/OpenDiablo2/sound"
	"github.com/OpenDiablo2/OpenDiablo2/ui"
	"github.com/hajimehoshi/ebiten"
)

type labelItem struct {
	Label     *ui.Label
	IsHeading bool
	Available bool
}

// Credits represents the credits scene
type Credits struct {
	uiManager          *ui.Manager
	soundManager       *sound.Manager
	fileProvider       common.FileProvider
	sceneProvider      SceneProvider
	creditsBackground  *common.Sprite
	exitButton         *ui.Button
	creditsText        []string
	labels             []*labelItem
	cycleTime          float64
	cyclesTillNextLine int
	doneWithCredits    bool
}

// CreateCredits creates an instance of the credits scene
func CreateCredits(fileProvider common.FileProvider, sceneProvider SceneProvider, uiManager *ui.Manager, soundManager *sound.Manager) *Credits {
	result := &Credits{
		fileProvider:       fileProvider,
		uiManager:          uiManager,
		soundManager:       soundManager,
		sceneProvider:      sceneProvider,
		labels:             make([]*labelItem, 0),
		cycleTime:          0,
		doneWithCredits:    false,
		cyclesTillNextLine: 0,
	}
	return result
}

// Load is called to load the resources for the credits scene
func (v *Credits) Load() []func() {
	return []func(){
		func() {
			v.creditsBackground = v.fileProvider.LoadSprite(resourcepaths.CreditsBackground, palettedefs.Sky)
			v.creditsBackground.MoveTo(0, 0)
		},
		func() {
			v.exitButton = ui.CreateButton(ui.ButtonTypeMedium, v.fileProvider, common.TranslateString("#970"))
			v.exitButton.MoveTo(30, 550)
			v.exitButton.OnActivated(func() { v.onExitButtonClicked() })
			v.uiManager.AddWidget(v.exitButton)
		},
		func() {
			fileData, _ := common.Utf16BytesToString(v.fileProvider.LoadFile(resourcepaths.CreditsText)[2:])
			v.creditsText = strings.Split(fileData, "\r\n")
			for i := range v.creditsText {
				v.creditsText[i] = strings.Trim(v.creditsText[i], " ")
			}
		},
	}
}

// Unload unloads the data for the credits scene
func (v *Credits) Unload() {

}

// Render renders the credits scene
func (v *Credits) Render(screen *ebiten.Image) {
	v.creditsBackground.DrawSegments(screen, 4, 3, 0)
	for _, label := range v.labels {
		if label.Available {
			continue
		}
		label.Label.Draw(screen)
	}
}

const secondsPerCycle = float64(0.02)

// Update runs the update logic on the credits scene
func (v *Credits) Update(tickTime float64) {
	v.cycleTime += tickTime
	for v.cycleTime >= secondsPerCycle {
		v.cycleTime -= secondsPerCycle
		v.cyclesTillNextLine--
		if !v.doneWithCredits && v.cyclesTillNextLine <= 0 {
			v.addNextItem()
		}

		for _, label := range v.labels {
			if label.Available {
				continue
			}
			if label.Label.Y-1 < -15 {
				label.Available = true
				continue
			}
			label.Label.Y--
		}
	}
}

func (v *Credits) onExitButtonClicked() {
	mainMenu := CreateMainMenu(v.fileProvider, v.sceneProvider, v.uiManager, v.soundManager)
	mainMenu.ShowTrademarkScreen = false
	v.sceneProvider.SetNextScene(mainMenu)
}

func (v *Credits) addNextItem() {
	if len(v.creditsText) == 0 {
		v.doneWithCredits = true
		return
	}

	text := v.creditsText[0]
	v.creditsText = v.creditsText[1:]
	if len(text) == 0 {
		v.cyclesTillNextLine = 18
		return
	}
	isHeading := text[0] == '*'
	isNextHeading := len(v.creditsText) > 0 && len(v.creditsText[0]) > 0 && v.creditsText[0][0] == '*'
	isNextSpace := len(v.creditsText) > 0 && len(v.creditsText[0]) == 0
	var label = v.getNewFontLabel(isHeading)
	if isHeading {
		label.SetText(text[1:])
	} else {
		label.SetText(text)
	}
	width, _ := label.GetSize()
	isDoubled := false
	if !isHeading && !isNextHeading && !isNextSpace {
		isDoubled = true

		// Gotta go side by side
		label.MoveTo(390-int(width), 605)

		text2 := v.creditsText[0]
		v.creditsText = v.creditsText[1:]

		isNextHeading = len(v.creditsText) > 0 && len(v.creditsText[0]) > 0 && v.creditsText[0][0] == '*'
		label2 := v.getNewFontLabel(isHeading)
		label2.SetText(text2)

		label2.MoveTo(410, 605)
	} else {
		label.MoveTo(400-int(width/2), 605)
	}

	if isHeading && isNextHeading {
		v.cyclesTillNextLine = 40
	} else if isNextHeading {
		if isDoubled {
			v.cyclesTillNextLine = 40
		} else {
			v.cyclesTillNextLine = 70
		}
	} else if isHeading {
		v.cyclesTillNextLine = 40
	} else {
		v.cyclesTillNextLine = 18
	}
}

func (v *Credits) getNewFontLabel(isHeading bool) *ui.Label {
	for _, label := range v.labels {
		if label.Available {
			label.Available = false
			if isHeading {
				label.Label.Color = color.RGBA{255, 88, 82, 255}
			} else {
				label.Label.Color = color.RGBA{198, 178, 150, 255}
			}
			return label.Label
		}
	}

	newLabelItem := &labelItem{
		Available: false,
		IsHeading: isHeading,
		Label:     ui.CreateLabel(v.fileProvider, resourcepaths.FontFormal10, palettedefs.Sky),
	}

	if isHeading {
		newLabelItem.Label.Color = color.RGBA{255, 88, 82, 255}
	} else {
		newLabelItem.Label.Color = color.RGBA{198, 178, 150, 255}

	}

	v.labels = append(v.labels, newLabelItem)
	return newLabelItem.Label
}
